# 1. Compiled Vs Interpretted

**Compiled** :
* Preprocess the instructions and create an executable which is ready to execute.

* Faster. Important when speed of execution is critical

* Compilation step is platform dependent


**Interpretted**:

* Processes instructions line by line during execution. Don't have build step.

* NO PLATORM specific preprocessing.


-----------------------------------------


Python is using Hybrid Approach.

**Step 1** : High level instructions(*.py) -> Byte-code equivalent(*.pyc)
* Platform independent

**Step -2** : Byte-code Instruction -> Python Virtual Machine(PVM)->Interpret and execute


-------------------------------------------

Code blocks segragated using White space.



# 2. Storage and Access



* Python data is represented by **Objects**
* Every python Object has 3 things associated with it :

      1.  Value
      2.  Type
      3.  ID


* Python doen't store pointers. So we cannot use any memory address to access some value.

* Python object is stored in **Heap**

* Based on where object is stored on heap, its ID value is determined and its associated with that oject throughout life time.




In [ ]:
x = 100
y = 100.5
z = 100
print(x,y,z)
print(id(x),id(y),id(z))
print(type(x),type(y),type(z))


100 100.5 100
139839414816080 139838988579440 139839414816080
<class 'int'> <class 'float'> <class 'int'>


* Above code shows all objects are associated with same ID.

* When we say x=100, 100 is an object and x is label pointing to 100

   So when we say y=100, there is another label created that points to 100.

* IF we perfrom, x=1,y=1,z=1, then noone is pointing to 100 so object 100 will be deleted from heap.

In [ ]:
x = 100
y = x
print(x,y)
x = 5
print(x,y)

a = [1, 2, 3, 4]
b = a
print(a,b)

a[0] = 100
print(a,b)


100 100
5 100
[1, 2, 3, 4] [1, 2, 3, 4]
[100, 2, 3, 4] [100, 2, 3, 4]


In above example, as list is a mutable object, if we change after its creation, change happens at same place. But values(x,y) are not mutable. So when we change x, it startes refereing to another object.

# 3. Mutability

Object is said to be mutable, if it can be changed post its creation.


In [ ]:
print(a[1])
print(id(a))
print(id(a[1]))
a[1] = 200

print(a[1])
print(id(a))
print(id(a[1]))

100
132797270826880
132798598384976
200
132797270826880
132798598388176


* As per above example, ID of list remains same, even after change of values in list.

* ID of list elements changes with value change


In [ ]:
import sys
print(sys.getrefcount(1))

17106


* There are 17106 references to Object "1"

# 4. Identifiers

* Same as C - Start with alphabet,digits allowed afterwards
*

In [ ]:
help("keywords")
help("and")
#1.11.35


Here is a list of the Python keywords.  Enter any keyword to get more help.

False               class               from                or
None                continue            global              pass
True                def                 if                  raise
and                 del                 import              return
as                  elif                in                  try
assert              else                is                  while
async               except              lambda              with
await               finally             nonlocal            yield
break               for                 not                 

Boolean operations
******************

   or_test  ::= and_test | or_test "or" and_test
   and_test ::= not_test | and_test "and" not_test
   not_test ::= comparison | "not" not_test

In the context of Boolean operations, and also when expressions are
used by control flow statements, the following values are interpreted
as false: "False

# 2 modes of operation : Interactive and Script

In [ ]:
x = "python"
dir(x)

['__add__',
 '__class__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getitem__',
 '__getnewargs__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__mod__',
 '__mul__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__rmod__',
 '__rmul__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 'capitalize',
 'casefold',
 'center',
 'count',
 'encode',
 'endswith',
 'expandtabs',
 'find',
 'format',
 'format_map',
 'index',
 'isalnum',
 'isalpha',
 'isascii',
 'isdecimal',
 'isdigit',
 'isidentifier',
 'islower',
 'isnumeric',
 'isprintable',
 'isspace',
 'istitle',
 'isupper',
 'join',
 'ljust',
 'lower',
 'lstrip',
 'maketrans',
 'partition',
 'removeprefix',
 'removesuffix',
 'replace',
 'rfind',
 'rindex',
 'rjust',
 'rpartition',
 'rsplit',
 'rstrip',
 'split',
 'splitlines',
 'startswith',
 'strip',
 'swapcase',


In [ ]:
print(x)
x.upper()


python


'python'

In [ ]:

x = []
bool(x)

False

In [ ]:
#Multi-line string
x = """ Rte_Call_VDM_DtcM_Impl_Api_CS_getDTCStatusByte(DTC_EVENT_ID_0xE40092, &status);
	bEventTested_0xE40092 = DiagnosticDtc_GetEventTested(status);
	bEventFailed_0xE40092 = DiagnosticDtc_GetEventFailed(status);
"""
print(x)

 Rte_Call_VDM_DtcM_Impl_Api_CS_getDTCStatusByte(DTC_EVENT_ID_0xE40092, &status);
	bEventTested_0xE40092 = DiagnosticDtc_GetEventTested(status);
	bEventFailed_0xE40092 = DiagnosticDtc_GetEventFailed(status);		



# String Operations
String is :
 Immutable
 Iterable


In [ ]:
s = "Wipro Technologies"
print(s[3])
print(s[-3])
print(s[0])
print(s[1])
print(s[-1])
print(len(s))

r
i
W
i
s
18


In [ ]:
#slicing
#[start,end,step]
print(s[::1])
print(s[6:9:1]) #9 is upto and non-inclusive
print(s[::-1])


Wipro Technologies
Tec
seigolonhceT orpiW
